modules/standard/Heap.chpl

```diff
@@ -0,0 +1,420 @@
+
+  A heap is a specialized tree-based data structure
+  that supports extracting the maximal or the minimal element quickly,
+  which can serve as a priority queue.
+
+  * Both `push` and `pop` operations are O(lgN).
+  * Querying the top element is O(1).
+  * Initialization from an array is O(N).
+*/
+module Heap {
+  import ChapelLocks;
+  private use HaltWrappers;
+  private use List;
+  private use IO;
+
+  public use Sort only defaultComparator, DefaultComparator,
+                       reverseComparator, ReverseComparator;
+  private use Sort;
+
+  // The locker is borrowed from List.chpl
+  // 
+  // We can change the lock type later. Use a spinlock for now, even if it
+  // is suboptimal in cases where long critical sections have high
+  // contention (IE, lots of tasks trying to call toArray on the heap
+  // or any operation that is O(n)).
+  //
+  pragma "no doc"
+  type _lockType = ChapelLocks.chpl_LocalSpinlock;
+
+  //
+  // Use a wrapper class to let heap methods have a const ref receiver even
+  // when `parSafe` is `true` and the heap lock is used.
+  //
+  pragma "no doc"
+  class _LockWrapper {
+    var lock$ = new _lockType();
+
+    inline proc lock() {
+      lock$.lock();
+    }
+
+    inline proc unlock() {
+      lock$.unlock();
+    }
+  }
+
+  pragma "no doc"
+  proc _checkType(type eltType) {
+    //NOTE: This is borrowed from List.chpl
+    if isGenericType(eltType) {
+      compilerWarning("creating a heap with element type " +
+                      eltType:string);
+      if isClassType(eltType) && !isGenericType(borrowed eltType) {
+        compilerWarning("which now means class type with generic management");
+      }
+      compilerError("heap element type cannot currently be generic");
+      // In the future we might support it if the list is not default-inited
+    }
+  }
+    
+  record heap {
+
+    /* The type of the elements contained in this heap. */
+    type eltType;
+
+    /* If `true`, this heap will perform parallel safe operations. */
+    param parSafe = false;
+
+    /*
+      Comparator record that defines how the
+      data is compared. The greatest element will be on the top.
+    */
+    var comparator: record;
+
+    pragma "no doc"
+    var _lock$ = if parSafe then new _LockWrapper() else none;
+
+    /*
+      Use a list to store elements.
+    */
+    pragma "no doc"
+    var _data: list(eltType);
+
+    /*
+      Build the heap from elements that have been stored, from bottom to top
+      in O(N)
+    */
+    pragma "no doc"
+    proc _commonInitFromIterable(iterable)
+    lifetime this < iterable {
+      _data = new list(eltType);
+      for x in iterable do
+        _data.append(x);
+      for i in 0 .. _data.size-1 by -1 {
+        _heapify_down(i);
+      }
+    }
+
+    /*
+      Initializes an empty heap.
+
+      :arg eltType: The type of the elements
+
+      :arg comparator: `DefaultComparator` makes max-heap and `ReverseCompartor` makes a min-heap
```

Comment:
The description of the argument should state explicitly that the comparator passed in will provide the heap's definition of order.  The examples are good, but should refer to the comparator objects `defaultComparator` and `reverseComparator`, not the names of their types.

Reply:
There is  description on `var comparator: record`. I just realized that maybe it's better to move these examples there and  keep something like `:arg comparator: the comparator used to compare elements` here

Reply:
https://github.com/chapel-lang/chapel/pull/15997/commits/b1cc51c5baaa5b4feed46431d5ae75c4368f95bd I commited some change here